File: android/src/main/java/com/lightningkite/khrysalis/observables/binding/DateButton.binding.actual.kt
package com.lightningkite.khrysalis.observables.binding

import com.lightningkite.khrysalis.observables.*
import com.lightningkite.khrysalis.rx.removed
import com.lightningkite.khrysalis.rx.until
import com.lightningkite.khrysalis.time.*
import com.lightningkite.khrysalis.views.android.DateButton
import com.lightningkite.khrysalis.views.android.TimeButton
import java.util.*

/**
 *
 * Binds the date buttons internal date to the ObservableProperty.
 * When the date button updates it will update the observable and vice versa.
 *
 *  Example
 *  val date = StandardObservableProperty(Date())
 *  dateButton.bind(date)
 *
 */

fun DateButton.bind(date: MutableObservableProperty<Date>) {
    date.subscribeBy { it ->
        this.date = it

    }.until(this.removed)
    this.onDateEntered.subscribe { it ->
        date.value = it
    }.until(this.removed)
}


/**
 *
 * Binds the time buttons internal time to the ObservableProperty<Date>.
 * When the time button updates it will update the observables time and vice versa.
 * Also you can specify the minute interval the user can choose from. All other time options
 * outside that interval will be greyed out.
 *
 *  Example
 *  val date = StandardObservableProperty(Date())
 *  timeButton.bind(date, 15)
 *
 */
fun TimeButton.bind(date: MutableObservableProperty<Date>, minuteInterval: Int = 1) {
    this.minuteInterval = minuteInterval
    date.subscribeBy { it ->
        this.date = it
    }.until(this.removed)
    this.onDateEntered.subscribe { it ->
        date.value = it
    }.until(this.removed)
}


/**
 *
 * Binds the date buttons internal time to the ObservableProperty<DateAlone>.
 * When the date button updates it will update the observables date and vice versa.
 *
 *  Example
 *  val dateAlone = StandardObservableProperty(Date().dateAlone)
 *  timeButton.bind(dateAlone)
 *
 */
fun DateButton.bindDateAlone(date: MutableObservableProperty<DateAlone>) {
    date.subscribeBy { it ->
        this.date = dateFrom(it, Date().timeAlone)
    }.until(this.removed)
    this.onDateEntered.subscribe { it ->
        date.value = it.dateAlone
    }.until(this.removed)
}


/**
 *
 * Binds the time buttons internal time to the ObservableProperty<TimeAlone>.
 * When the time button updates it will update the observables time and vice versa.
 * Also you can specify the minute interval the user can choose from. All other time options
 * outside that interval will be greyed out.
 *
 *  Example
 *  val time = StandardObservableProperty(Date().timeAlone)
 *  timeButton.bind(time, 15)
 *
 */
fun TimeButton.bindTimeAlone(date: MutableObservableProperty<TimeAlone>, minuteInterval: Int = 1) {
    this.minuteInterval = minuteInterval
    date.subscribeBy { it ->
        this.date = dateFrom(Date().dateAlone, it)
    }.until(this.removed)
    this.onDateEntered.subscribe { it ->
        date.value = it.timeAlone
    }.until(this.removed)
}

File: ios/Khrysalis/src/observables/binding/DateButton.binding.actual.swift
//Stub file made with Khrysalis 2 (by Lightning Kite)
import Foundation


//--- DateButton.bind(MutableObservableProperty<Date>)
public extension DateButton {
    func bind(_ observable: MutableObservableProperty<Date>) -> Void {
        self.date = observable.value
        observable.subscribeBy { ( value) in
            if self.date != value {
                self.date = value
            }
        }.until(self.removed)
        self.onDateEntered.subscribeBy { value in
            if observable.value != value {
                observable.value = value
            }
        }.until(self.removed)
    }
    func bind(date: MutableObservableProperty<Date>) -> Void {
        return bind(date)
    }
}

//--- TimeButton.bind(MutableObservableProperty<Date>, Int)
public extension TimeButton {
    func bind(_ observable: MutableObservableProperty<Date>, _ minuteInterval: Int = 1) -> Void {
        self.minuteInterval = Int(minuteInterval)
        self.date = observable.value
        observable.subscribeBy { ( value) in
            if self.date != value {
                self.date = value
            }
        }.until(self.removed)
        self.onDateEntered.subscribeBy { value in
            if observable.value != value {
                observable.value = value
            }
        }.until(self.removed)
    }
    func bind(date: MutableObservableProperty<Date>, minuteInterval: Int = 1) -> Void {
        return bind(date, minuteInterval)
    }
}

//--- DateButton.bindDateAlone(MutableObservableProperty<DateAlone>)
public extension DateButton {
    func bindDateAlone(_ observable: MutableObservableProperty<DateAlone>) -> Void {
        observable.subscribeBy { ( value) in
            if self.date.dateAlone != value {
                self.date = dateFrom(dateAlone: value, timeAlone: Date().timeAlone)
            }
        }.until(self.removed)
        self.onDateEntered.subscribeBy { value in
            let newValue = self.date.dateAlone
            if observable.value != newValue {
                observable.value = newValue
            }
        }.until(self.removed)
    }
    func bindDateAlone(date: MutableObservableProperty<DateAlone>) -> Void {
        return bindDateAlone(date)
    }
}

//--- TimeButton.bindTimeAlone(MutableObservableProperty<TimeAlone>, Int)
public extension TimeButton {
    func bindTimeAlone(_ observable: MutableObservableProperty<TimeAlone>, _ minuteInterval: Int = 1) -> Void {
        self.minuteInterval = Int(minuteInterval)
        observable.subscribeBy { ( value) in
            if self.date.timeAlone != value {
                self.date = dateFrom(dateAlone: Date().dateAlone, timeAlone: value)
            }
        }.until(self.removed)
        self.onDateEntered.subscribeBy { value in
            let newValue = self.date.timeAlone
            if observable.value != newValue {
                observable.value = newValue
            }
        }.until(self.removed)
    }
    func bindTimeAlone(date: MutableObservableProperty<TimeAlone>, minuteInterval: Int = 1) -> Void {
        return bindTimeAlone(date, minuteInterval)
    }
}
